File: internal/parse/parse.go
// Package parse provides shared parsing utilities for sizes, seed times, and numbers.
// These are used by the stats fetching layer and the history recorder.
package parse

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var sizeRe = regexp.MustCompile(`(?i)([\d.]+)\s*(B|[KMGTP]i?B)`)

var sizeFactors = map[string]float64{
	"b":   1.0 / (1024 * 1024 * 1024),
	"kib": 1.0 / (1024 * 1024),
	"mib": 1.0 / 1024,
	"gib": 1.0,
	"tib": 1024.0,
	"pib": 1024.0 * 1024.0,
	// Decimal-labelled units (TBDev-family sites render "1.005 TB"): tracker
	// software near-universally computes 1024-based sizes but mislabels them,
	// so these map to the same factors as their -iB counterparts.
	"kb": 1.0 / (1024 * 1024),
	"mb": 1.0 / 1024,
	"gb": 1.0,
	"tb": 1024.0,
	"pb": 1024.0 * 1024.0,
}

// SizeToGiB parses a human-readable size string (e.g. "3.14 TiB") into GiB.
// Returns nil when the input is empty or unparseable.
func SizeToGiB(s string) *float64 {
	m := sizeRe.FindStringSubmatch(s)
	if m == nil {
		return nil
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return nil
	}
	f, ok := sizeFactors[strings.ToLower(m[2])]
	if !ok {
		return nil
	}
	result := v * f
	return &result
}

// SeedTimeToSeconds parses a seed time string (e.g. "3M 6D 22h 23m 13s") to seconds.
// Returns nil for empty/invalid input.
// Also accepts a plain integer/float as raw seconds.
func SeedTimeToSeconds(s string) *float64 {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	// Plain numeric — treat as seconds directly
	if v, err := strconv.ParseFloat(s, 64); err == nil {
		if v > 0 {
			return &v
		}
		return nil
	}
	grp := func(pattern string, flags ...string) float64 {
		p := pattern
		if len(flags) > 0 && flags[0] == "i" {
			p = "(?i)" + p
		}
		re := regexp.MustCompile(p)
		m := re.FindStringSubmatch(s)
		if m == nil {
			return 0
		}
		v, _ := strconv.ParseFloat(m[1], 64)
		return v
	}
	total := grp(`([\d.]+)\s*Y`, "i")*365.25*86400 +
		grp(`([\d.]+)\s*M`)*30.44*86400 + // uppercase M only
		grp(`([\d.]+)\s*W`, "i")*7*86400 +
		grp(`([\d.]+)\s*D`, "i")*86400 +
		grp(`([\d.]+)\s*h`)*3600 +
		grp(`([\d.]+)\s*m`)*60 + // lowercase m only
		grp(`([\d.]+)\s*s`)

	if total <= 0 {
		return nil
	}
	return &total
}

// AnyFloat converts any numeric JSON value to float64.
func AnyFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	case string:
		f, _ := strconv.ParseFloat(strings.ReplaceAll(n, ",", ""), 64)
		return f
	}
	return 0
}

// BytesToSize converts a raw byte count (as returned by the Gazelle API) to a
// human-readable size string with exactly 2 decimal places, e.g. "25.52 GiB".
func BytesToSize(b int64) string {
	if b <= 0 {
		return "0.00 B"
	}
	const (
		kib = int64(1024)
		mib = kib * 1024
		gib = mib * 1024
		tib = gib * 1024
		pib = tib * 1024
	)
	switch {
	case b >= pib:
		return fmt.Sprintf("%.2f PiB", float64(b)/float64(pib))
	case b >= tib:
		return fmt.Sprintf("%.2f TiB", float64(b)/float64(tib))
	case b >= gib:
		return fmt.Sprintf("%.2f GiB", float64(b)/float64(gib))
	case b >= mib:
		return fmt.Sprintf("%.2f MiB", float64(b)/float64(mib))
	case b >= kib:
		return fmt.Sprintf("%.2f KiB", float64(b)/float64(kib))
	default:
		return fmt.Sprintf("%d B", b)
	}
}

// NormalizeSeedSize re-formats a size string to exactly 2 decimal places.
// Handles scraped values like "1.239 TiB" → "1.24 TiB".
// Leaves strings that don't match a known unit pattern unchanged.
func NormalizeSeedSize(s string) string {
	m := sizeRe.FindStringSubmatch(s)
	if m == nil {
		return s
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return s
	}
	return fmt.Sprintf("%.2f %s", v, m[2])
}

// FormatSeedTime formats seconds into a human string like "1Y 2M 3D 4h 5m 6s".
func FormatSeedTime(totalSec float64) string {
	t := int64(totalSec)
	if t <= 0 {
		return "0s"
	}
	steps := []struct {
		sec int64
		u   string
	}{
		{31536000, "Y"}, {2592000, "M"}, {604800, "W"},
		{86400, "D"}, {3600, "h"}, {60, "m"}, {1, "s"},
	}
	var parts []string
	for _, s := range steps {
		v := t / s.sec
		t -= v * s.sec
		if v > 0 {
			parts = append(parts, fmt.Sprintf("%d%s", v, s.u))
		}
	}
	if len(parts) == 0 {
		return "0s"
	}
	return strings.Join(parts, " ")
}

File: web/src/utils/parse.ts
// utils/parse.ts — parsing helpers for sizes, times, dates

/** Parse a size string (e.g. "3.14 TiB") → GiB as number, or null */
export function parseSize(str: unknown): number | null {
  if (!str) return null;
  const m = String(str).match(/([\d.]+)\s*(B|[KMGTP]i?B)/i);
  if (!m) return null;
  // Decimal-labelled units (kb/mb/gb/tb) map to the binary factors — tracker
  // software computes 1024-based sizes but often mislabels them.
  const factors: Record<string, number> = {
    b: 1 / (1024 ** 3), kib: 1 / (1024 ** 2), mib: 1 / 1024,
    gib: 1, tib: 1024, pib: 1048576,
    kb: 1 / (1024 ** 2), mb: 1 / 1024, gb: 1, tb: 1024, pb: 1048576,
  };
  return parseFloat(m[1]) * (factors[m[2].toLowerCase()] ?? 1);
}

/**
 * Parse seed time string → total seconds.
 * Supports: Y M W D h m s
 * Case rules:
 *   Y/y = years, M = months (UPPERCASE only), W/w = weeks, D/d = days,
 *   h = hours, m = minutes (lowercase only), s = seconds
 * Also accepts a plain integer/float as raw seconds.
 */
export function parseSeedTime(str: unknown): number | null {
  if (str == null || str === '') return null;
  const s = String(str).trim();
  if (!s) return null;
  // Plain number → raw seconds
  if (/^\d+(\.\d+)?$/.test(s)) return Math.round(parseFloat(s));

  let total = 0, found = false;
  const grp = (pattern: RegExp): number => {
    const m = s.match(pattern);
    if (!m) return 0;
    found = true;
    return parseFloat(m[1]);
  };
  total += grp(/(\d+(?:\.\d+)?)\s*[Yy](?![a-zA-Z])/) * 365 * 86400;
  total += grp(/(\d+(?:\.\d+)?)\s*M(?![a-zA-Z])/)    * 30  * 86400; // uppercase M
  total += grp(/(\d+(?:\.\d+)?)\s*[Ww](?![a-zA-Z])/) * 7   * 86400;
  total += grp(/(\d+(?:\.\d+)?)\s*[Dd](?![a-zA-Z])/) * 86400;
  total += grp(/(\d+(?:\.\d+)?)\s*h(?![a-zA-Z])/)    * 3600;
  total += grp(/(\d+(?:\.\d+)?)\s*m(?![a-zA-Z])/)    * 60;  // lowercase m
  total += grp(/(\d+(?:\.\d+)?)\s*s(?![a-zA-Z])/);
  return found ? Math.round(total) : null;
}

/** Calculate account age in whole days from a YYYY-MM-DD string */
export function memberDays(d: string): number | null {
  if (!d) return null;
  const j = new Date(d);
  if (isNaN(j.getTime())) return null;
  return Math.floor((Date.now() - j.getTime()) / 86_400_000);
}

/** Format account age in days to "1Y 2M 3W 4D" */
export function memberDur(d: string): string {
  const days = memberDays(d);
  if (days === null || days < 0) return '—';
  if (days === 0) return '0D';
  const Y = Math.floor(days / 365), r1 = days % 365;
  const M = Math.floor(r1 / 30),   r2 = r1 % 30;
  const W = Math.floor(r2 / 7),    D = r2 % 7;
  return ([[Y,'Y'],[M,'M'],[W,'W'],[D,'D']] as [number,string][])
    .filter(([v]) => v)
    .map(([v, u]) => `${v}${u}`)
    .join(' ') || '0D';
}

/**
 * Parse account-age target: plain number (days) or "Y M W D" format.
 * Returns total days.
 */
export function parseAgeDays(str: string): number | null {
  if (!str) return null;
  const s = String(str).trim();
  if (/^\d+$/.test(s)) return parseInt(s);
  let total = 0;
  const yr = s.match(/(\d+)\s*Y/i); if (yr) total += parseInt(yr[1]) * 365;
  const mo = s.match(/(\d+)\s*M/);  if (mo) total += parseInt(mo[1]) * 30;
  const wk = s.match(/(\d+)\s*W/i); if (wk) total += parseInt(wk[1]) * 7;
  const dd = s.match(/(\d+)\s*D/i); if (dd) total += parseInt(dd[1]);
  return total > 0 ? total : null;
}

/** Build a favicon URL from a tracker URL */
export function getFaviconUrl(trackerUrl: string): string {
  try {
    const u = new URL(trackerUrl);
    return `${u.protocol}//${u.host}/favicon.ico`;
  } catch { return ''; }
}
